### ISSUE TITLE
Support the new Tesla command protocol

### ISSUE THREAD

**Issue by gshaviv** [jonasman/TeslaSwift#160]
As the old command REST API is being deprecated, the new way of sending commands is via the new Tesla Command Protocol using the signed_command endpoint.

**Comment by gshaviv** [comment]
Indeed. According to the fleet API docs, the old REST api for sending commands will be deprecated and of the month and cease to work and starting Jan 24 sending commands will be done using the new command protocol and the new signed_command endpoint

**Comment by gshaviv** [comment]
Here is the link to the notification the old command API is phased out. https://developer.tesla.com/docs/fleet-api#announcements-amp-api-changelog

**Comment by jonasman** [comment]
the current API version already uses the new API.

**Comment by jonasman** [comment]
ok i think i found the new endpoint: `signed_command`

**Comment by NicolasJott** [comment]
is this related to the response i've been getting when trying to send vehicle commands?

`RESPONSE BODY: {"response":{"result":false,"reason":"unsigned_cmds_hardlocked"}}`

**Comment by gshaviv** [comment]
Yep. The new way to send commands is to sign them with your private key so the car can verify them with your public key

**Comment by jonasman** [comment]
I wonder if we shall use the proxy?
this new API is a bit confusing to understand what is needed. I will try to see if we can use `signed_command`

**Comment by jonasman** [comment]
tried to have a look at signed_command, but im not sure how to build the message to send the command, probably need to base64 encode the protbuf msg somehow.

**Comment by gshaviv** [comment]
I saw in the proxy repository someone from Tesla wrote they are working on publishing docs for this. Hopefully will be available soon

**Comment by gshaviv** [comment]
Tesla added a doc explaining how messages are encoded to the proxy repo, maybe now we can implement this in the package:

https://github.com/teslamotors/vehicle-command/commit/d7a3e30f4f631b36760c61bc980157ea6de6d2bd

**Comment by craigrigdon** [comment]
any insight on encoding the protbuf msg to be send via signed_command yet?

**Comment by gshaviv** [comment]
@Urkman Sorry, haven't had time to try it. Kinda busy and overwhelmed with other stuff at the moment.

**Comment by jonasman** [comment]
we need the pairing to work as well, not just sending messages. I recently installed Charge HQ and they have the new pairing working.

1. First the user logs in with Tesla credentials
2. Then the user needs to open a link that will in turn open the Tesla app
3. The Tesla app asks to install some virtual key (for me it said it was not needed because my model S is old)
4. After this virtual key is installed, i think the permissions are done

**Comment by jonasman** [comment]
i found how to do #2 will be pushing the code soon
Tesla has updated the docs with more info

**Comment by jonasman** [comment]
tag 9.2.0 includes a helper to send the public key to the vehicle

**Comment by jonasman** [comment]
@gshaviv is the API now working fine for you? should we close this issue?

**Comment by gshaviv** [comment]
Vehicle commands are still not supported. Has any change been attempted to support the signed command?

**Comment by jonasman** [comment]
unfortunately i dont have a model 3, so im not able to test. 
Do you mean that after we send the pub key we need to send all commands via the new signed_command API?

**Comment by gshaviv** [comment]
@jonasman exactly. The old way of sending commands no longer works. Only signed commands work.

**Comment by jonasman** [comment]
@gshaviv any idea on the api of that request?

**Comment by Urkman** [comment]
Anyone managed to sign a command using swift and send it to the API?
https://developer.tesla.com/docs/fleet-api#signed_command

**Comment by jonasman** [comment]
@Urkman as far as i know, nobody is able to do it without using HTTP proxys yet.
